### ISSUE TITLE
Missing Mach-O / Apple Platform Types

### ISSUE THREAD

**Issue by rickmark** [Vector35/binaryninja-api#4048]
Primary Reference: https://github.com/apple/darwin-xnu/blob/2ff845c2e033bd0ff64b5b6aa6063a1f8f65aa32/EXTERNAL_HEADERS/mach-o/loader.h#L4
Creating a meta-issue of what types are missing from Mach-O parsing:

## LC_LOAD_DYLINKER
```
struct dylinker_command {
	uint32_t	cmd;		/* LC_ID_DYLINKER, LC_LOAD_DYLINKER or
					   LC_DYLD_ENVIRONMENT */
	uint32_t	cmdsize;	/* includes pathname string */
	union lc_str    name;		/* dynamic linker's path name , offset into command where string appears, should be after struct.  create x-ref */
};
// Immediate follow with C-String of the name of the linker
```

## LC_BUILD_VERSION
```
struct build_version_command {
    uint32_t	cmd;		/* LC_BUILD_VERSION */
    uint32_t	cmdsize;	/* sizeof(struct build_version_command) plus */
                                /* ntools * sizeof(struct build_tool_version) */
    uint32_t	platform;	/* platform */
    uint32_t	minos;		/* X.Y.Z is encoded in nibbles xxxx.yy.zz */
    uint32_t	sdk;		/* X.Y.Z is encoded in nibbles xxxx.yy.zz */
    uint32_t	ntools;		/* number of tool entries following this */
};

struct build_tool_version {
    uint32_t	tool;		/* enum for the tool */
    uint32_t	version;	/* version number of the tool */
};
```

## LC_SOURCE_VERSION
```
struct source_version_command {
    uint32_t  cmd;	/* LC_SOURCE_VERSION */
    uint32_t  cmdsize;	/* 16 */
    uint64_t  version;	/* A.B.C.D.E packed as a24.b10.c10.d10.e10 */
};
```

## LC_MAIN
```
struct entry_point_command {
    uint32_t  cmd;	/* LC_MAIN only used in MH_EXECUTE filetypes */
    uint32_t  cmdsize;	/* 24 */
    uint64_t  entryoff;	/* file (__TEXT) offset of main() */
    uint64_t  stacksize;/* if not zero, initial stack size */
};
```

## LC_FILESET
```
struct fileset_entry_command {
    uint32_t        cmd;        /* LC_FILESET_ENTRY */
    uint32_t        cmdsize;    /* includes id string */
    uint64_t        vmaddr;     /* memory address of the dylib */
    uint64_t        fileoff;    /* file offset of the dylib */
    union lc_str    entry_id;   /* contained entry id, offset from start of command to string with name, should immediately follow, xref to string */
    uint32_t        reserved;   /* entry_id is 32-bits long, so this is the reserved padding */
};
// Immediately follow with C-String with the fileset name
```

## file_type_t
Modify `file_type_t` to be:

```
enum file_type_t : uint32_t
{
    MH_OBJECT = 0x1,
    MH_EXECUTE = 0x2,
    MH_FVMLIB = 0x3,
    MH_CORE = 0x4,
    MH_PRELOAD = 0x5,
    MH_DYLIB = 0x6,
    MH_DYLINKER = 0x7,
    MH_BUNDLE = 0x8,
    MH_DYLIB_STUB = 0x9,
    MH_DSYM = 0xa,
    MH_KEXT_BUNDLE = 0xb,
    MH_FILESET = 0xc,
};
```

## load_command_type_t
Modify `load_command_type_t` to be:

```
enum load_command_type_t : uint32_t
{
    LC_REQ_DYLD = 0x80000000,
    LC_SEGMENT = 0x1,
    LC_SYMTAB = 0x2,
    LC_SYMSEG = 0x3,
    LC_THREAD = 0x4,
    LC_UNIXTHREAD = 0x5,
    LC_LOADFVMLIB = 0x6,
    LC_IDFVMLIB = 0x7,
    LC_IDENT = 0x8,
    LC_FVMFILE = 0x9,
    LC_PREPAGE = 0xa,
    LC_DYSYMTAB = 0xb,
    LC_LOAD_DYLIB = 0xc,
    LC_ID_DYLIB = 0xd,
    LC_LOAD_DYLINKER = 0xe,
    LC_ID_DYLINKER = 0xf,
    LC_PREBOUND_DYLIB = 0x10,
    LC_ROUTINES = 0x11,
    LC_SUB_FRAMEWORK = 0x12,
    LC_SUB_UMBRELLA = 0x13,
    LC_SUB_CLIENT = 0x14,
    LC_SUB_LIBRARY = 0x15,
    LC_TWOLEVEL_HINTS = 0x16,
    LC_PREBIND_CKSUM = 0x17,
    LC_LOAD_WEAK_DYLIB = 0x80000018,
    LC_SEGMENT_64 = 0x19,
    LC_ROUTINES_64 = 0x1a,
    LC_UUID = 0x1b,
    LC_RPATH = 0x8000001c,
    LC_CODE_SIGNATURE = 0x1d,
    LC_SEGMENT_SPLIT_INFO = 0x1e,
    LC_REEXPORT_DYLIB = 0x8000001f,
    LC_LAZY_LOAD_DYLIB = 0x20,
    LC_ENCRYPTION_INFO = 0x21,
    LC_DYLD_INFO = 0x22,
    LC_DYLD_INFO_ONLY = 0x80000022,
    LC_LOAD_UPWARD_DYLIB = 0x80000023,
    LC_VERSION_MIN_MACOSX = 0x24,
    LC_VERSION_MIN_IPHONEOS = 0x25,
    LC_FUNCTION_STARTS = 0x26,
    LC_DYLD_ENVIRONMENT = 0x27,
    LC_MAIN = 0x80000028,
    LC_DATA_IN_CODE = 0x29,
    LC_SOURCE_VERSION = 0x2a,
    LC_DYLIB_CODE_SIGN_DRS = 0x2b,
    LC_ENCRYPTION_INFO_64 = 0x2c,
    LC_LINKER_OPTION = 0x2d,
    LC_LINKER_OPTIMIZATION_HINT = 0x2e,
    LC_VERSION_MIN_TVOS = 0x2f,
    LC_VERSION_MIN_WATCHOS = 0x30,
    LC_NOTE = 0x31,
    LC_BUILD_VERSION = 0x32,
    LC_DYLD_EXPORTS_TRIE = 0x80000033,
    LC_DYLD_CHAINED_FIXUPS = 0x80000034,
    LC_FILESET_ENTRY = 0x800000035
};
```

## nlist / nlist64
Add `nlist64` for 64bit and `nlist` for 32 bit (https://github.com/apple/darwin-xnu/blob/2ff845c2e033bd0ff64b5b6aa6063a1f8f65aa32/EXTERNAL_HEADERS/mach-o/nlist.h):

```
struct nlist {
	union {
		uint32_t n_strx;	/* index into the string table */
	} n_un;
	uint8_t n_type;		/* type flag, see below */
	uint8_t n_sect;		/* section number or NO_SECT */
	int16_t n_desc;		/* see <mach-o/stab.h> */
	uint32_t n_value;	/* value of this symbol (or stab offset) */
};

struct nlist_64 {
    union {
        uint32_t  n_strx; /* index into the string table */
    } n_un;
    uint8_t n_type;        /* type flag, see below */
    uint8_t n_sect;        /* section number or NO_SECT */
    uint16_t n_desc;       /* see <mach-o/stab.h> */
    uint64_t n_value;      /* value of this symbol (or stab offset) */
};
```

**Comment by plafosse** [comment]
So many of these do exist in our `machoview.h` Additionally unless there is a compelling reason to have I don't believe we need to add these.

**Comment by fuzyll** [comment]
To be clear, this is the file that @plafosse is referring to: https://github.com/Vector35/view-macho/blame/main/machoview.h. It's a part of our open-sourced Mach-O code that I mentioned in my reply to #3815. This wasn't available at the time this issue (or #4046) were created.

If you're seeing issues where these things aren't being applied correctly, definitely let us know.

For completeness, though, the following things @rickmark requested are not a part of that file:

* `struct source_version_command`
* `struct entry_point_command`
* `struct fileset_entry_command`
* `enum file_type_t`
* `enum load_command_type_t`

Some of these, like the last one, are handled differently in the C++ source file instead: https://github.com/Vector35/view-macho/blob/5a99c92581fc255804aadbe11bf0ab3e3435e350/machoview.cpp#L30.

If you could help us understand what's missing that you need and why, that might help us make sure we add the right things to support your use-case better.
